chain/postgresdb/database/database.go

```diff
@@ -0,0 +1,334 @@
+	ErrUndefinedTable    = errors.New("undefined table")
+)
+
+// Config is the required properties to use the database.
+type Config struct {
+	User         string
+	Password     string
+	Host         string
+	Name         string
+	Schema       string
+	MaxIdleConns int
+	MaxOpenConns int
+	DisableTLS   bool
+}
+
+// ConfigFromDSN provides support for creating a config from a DSN.
+func ConfigFromDSN(dsn string) (Config, error) {
+	conf, err := url.Parse(dsn)
+	if err != nil {
+		return Config{}, err
+	}
+	query := conf.Query()
+
+	password, _ := conf.User.Password()
+	cfg := Config{
+		User:       conf.User.Username(),
+		Password:   password,
+		Host:       conf.Host,
+		Name:       conf.Path,
+		DisableTLS: true,
+	}
+
+	if query.Has("sslmode") {
+		sslMode := query.Get("sslmode")
+		switch sslMode {
+		//nolint:goconst // Having to add constants is overkill.
+		case "disable":
+			cfg.DisableTLS = true
+		case "required":
+			cfg.DisableTLS = false
+		default:
+			return Config{}, fmt.Errorf("unsupported ssl mode %q. Expected disable or required", sslMode)
+		}
+	}
+
+	cfg.MaxIdleConns = 2
+	if query.Has("max-idle") {
+		max := query.Get("max-idle")
+		m, err := strconv.Atoi(max)
+		if err != nil {
+			return Config{}, fmt.Errorf("expected number for max-idle, got err: %w", err)
+		}
+		cfg.MaxIdleConns = m
+	}
+
+	cfg.MaxOpenConns = 0
+	if query.Has("max-open") {
+		open := query.Get("max-open")
+		o, err := strconv.Atoi(open)
+		if err != nil {
+			return Config{}, fmt.Errorf("expected number for max-open, got err: %w", err)
+		}
+		cfg.MaxOpenConns = o
+	}
+
+	if query.Has("search_path") {
+		cfg.Schema = query.Get("search_path")
+	}
+
+	return cfg, nil
+}
+
+// Open knows how to open a database connection based on the configuration.
+// It also performs a health check to make sure the connection is healthy.
+//
+//nolint:gocritic // There is nothing wrong with using value semantics here.
+func Open(ctx context.Context, cfg Config) (*sqlx.DB, error) {
+	sslMode := "require"
+	if cfg.DisableTLS {
+		sslMode = "disable"
+	}
+
+	q := make(url.Values)
+	q.Set("sslmode", sslMode)
+	q.Set("timezone", "utc")
+	if cfg.Schema != "" {
+		q.Set("search_path", cfg.Schema)
+	}
+
+	u := url.URL{
+		Scheme:   "postgres",
+		User:     url.UserPassword(cfg.User, cfg.Password),
+		Host:     cfg.Host,
+		Path:     strings.ToLower(cfg.Name),
+		RawQuery: q.Encode(),
+	}
+
+	db, err := sqlx.Open("postgres", u.String())
+	if err != nil {
+		return nil, err
+	}
+	db.SetMaxIdleConns(cfg.MaxIdleConns)
+	db.SetMaxOpenConns(cfg.MaxOpenConns)
+
+	return db, StatusCheck(ctx, db)
+}
+
+// StatusCheck returns nil if it can successfully talk to the database. It
+// returns a non-nil error otherwise.
+func StatusCheck(ctx context.Context, db *sqlx.DB) error {
+	var pingError error
+	for attempts := 1; ; attempts++ {
+		pingError = db.Ping()
+		if pingError == nil {
+			break
+		}
+		time.Sleep(time.Duration(attempts) * 100 * time.Millisecond)
+		if ctx.Err() != nil {
```

Comment:
seems strange to check this inside the loop _and_ outside it - in some cases it avoids an extra `db.Ping()` - is it just to check for cancellations after the sleep?

Reply:
Good point. I'll rewrite this.

Reply:
You want to check in both places. Assume the loop passes but the context has also timed out. If you don't do the outside check, you may not repect the timeout. I would not change this on the semantics.